libcxx/include/math.h

```diff
@@ -307,53 +307,79 @@ long double    truncl(long double x);
 // back to C++ linkage before including these C++ headers.
 extern "C++" {
 
-#    ifdef fpclassify
-#      undef fpclassify
-#    endif
-
-#    ifdef signbit
-#      undef signbit
-#    endif
-
-#    ifdef isfinite
-#      undef isfinite
-#    endif
-
-#    ifdef isinf
-#      undef isinf
-#    endif
-
-#    ifdef isnan
-#      undef isnan
-#    endif
-
-#    ifdef isnormal
-#      undef isnormal
-#    endif
-
-#    ifdef isgreater
-#      undef isgreater
-#    endif
-
-#    ifdef isgreaterequal
-#      undef isgreaterequal
-#    endif
-
-#    ifdef isless
-#      undef isless
-#    endif
-
-#    ifdef islessequal
-#      undef islessequal
-#    endif
-
-#    ifdef islessgreater
-#      undef islessgreater
-#    endif
-
-#    ifdef isunordered
-#      undef isunordered
-#    endif
+// According to section 7.1.4 Use of library functions of the C standard, any
```

Comment:
IMO it would make more sense to undefine each macro where the corresponding libc++ function is implemented. For example in 

```c++
#undef cosh
inline _LIBCPP_HIDE_FROM_ABI float cosh(float __x) _NOEXCEPT { return __builtin_coshf(__x); }
// etc...
```

in `libcxx/include/__math/hyperbolic_functions.h`.

Reply:
I tried that in https://github.com/llvm/llvm-project/pull/94533/commits/745745c5df98cce4924f0fb8ef5bdffb1081953f but that lead to build errors, see https://github.com/llvm/llvm-project/pull/94533#issuecomment-2153159848.

I noticed that the particular `<cmath>` include that triggered that build error was removed by @philnik777 in https://github.com/llvm/llvm-project/commit/0dd8c0ddd0bef1753f7bcbd6709a8e99af507ed5 so it might worth trying it again and see if there are any remaining issues.

Reply:
I read that comment, but I still don't understand what's happening that would prevent the `#undef` from being located right before the definition of the libc++ function with the same name. It seems like we need exactly one canonical place where we `#undef foo` and then define `namespace std { int foo(); }`, and I don't see how this can lead to problems?

Reply:
@ldionne Consider
```c++
#include <__math/traits.h>
#include <math.h>
```
The `__math/traits.h` would try to `#undef` a macro that will only be defined when including `math.h`. Does that make it clearer?